_posts/2017-05-30-predicting-artist-career-stage-with-machine-learning-in-apache-spark.markdown

```diff
@@ -0,0 +1,181 @@
+       |  birthday,
+       |  max_price_in_auctions
+       |FROM
+       |  data.artists
+       |""".stripMargin
+  ).rdd.map { row =>
+    Features(
+      artistId        = row.getString(0),
+      careerStageGene = row.getDouble(1),
+      age             = DateTime.now.getYear - Try(DateTime.parse(row.getString(2)).getYear).getOrElse(DateTime.now.getYear),
+      maxPrice        = row.getDouble(3)
+    )
+  }
+}
+```
+
+Features are packed in a vector to become usable by the built-in Spark functions. We're not using the manually entered career stage gene as input, so it's not part of the feature vector.
+
+```scala
+import org.apache.spark.mllib.linalg._
+
+case class Features(
+   ...
+) = {
+  def vector: Vector = {
+    Vectors.dense(
+      age.toDouble,
+      maxPrice
+    )
+  }
+
+```
+
+### Normalizing and Weighing Features
+
+Since fetching a million dollars at auction isn't a million times more important than an artist's age, we must normalize values. We create a scaler to be used in both the training and the prediction parts of the code.
+
+```scala
+import org.apache.spark.mllib.feature.StandardScaler
+
+val scaler = new StandardScaler().fit(features.map(f => f.vector))
+```
+
+### Preparing Training Points
+
+Our training data contains all the artists that have a manual value for a career stage gene. The input is the scaled vector of features and the output is the known career stage value.
+
+```scala
+val featuresForTraining = features.filter(feature => feature.careerStageGene > 0)
+val labeledPointsForTraining = featuresForTraining.map {
+  f => LabeledPoint(f.careerStageGene, scaler.transform(f.vector))
+}
+```
+
+### Training a Model
+
+We use a linear regression with stochastic gradient descent and an intercept (the value of career stage when the input features are zero) and train it. The optimizer options below are the default, but may need some experimenting depending on input data.
+
+```scala
+import org.apache.spark.mllib.regression._
+
+val model = new LinearRegressionWithSGD()
+model.optimizer.setNumIterations(100)
+model.optimizer.setStepSize(1.0)
+model.optimizer.setMiniBatchFraction(1.0)
+model.setIntercept(true)
+
+val trainedModel = model.run(labeledPointsForTraining)
+
+logInfo(s"weights: ${trainedModel.weights}, intercept: ${trainedModel.intercept}")
+```
+
+We can use the generated model to run predictions on known data and compare results using `RegressionMetrics`.
+
+```scala
+import org.apache.spark.mllib.evaluation._
+
+val predictionsAndObservations = labeledPointsForTraining.map { point =>
+  val prediction = model.predict(point.features)
+  (prediction, point.label)
+}
+
+val metrics = new RegressionMetrics(predictionsAndObservations)
+
+logInfo(s"mean squared error: ${metrics.meanSquaredError}")
+logInfo(s"r2=${metrics.r2}")
+logInfo(s"explained variance=${metrics.explainedVariance}")
+```
+
+The lower the MSE, the better and should be used to tune the optimizer above.
```

Comment:
you should use 'mean squared error (MSE)' in this line for clarity. The grammar is also strange in this sentence "the better and" is an odd conjunction. Lastly, we could clarify what the MSE represents and how it's used. This might be better:

"The lower the Mean Squared Error (MSE), the better the fit achieved by our linear regressor. The MSE is used to guide the gradient descent process during the learning process."